src/BreakpointViewer.cpp

```diff
@@ -438,278 +477,364 @@ void BreakpointViewer::setBreakState()
 	runState = false;
 }
 
-void BreakpointViewer::sync()
+void BreakpointViewer::stretchTable(BreakpointData::Type type)
 {
-	auto sa = ScopedAssign(userMode, false);
-	keepUnusedItems();
-
-	int bpRow = bpTableWidget->rowCount();
-	int wpRow = wpTableWidget->rowCount();
-	int cnRow = cnTableWidget->rowCount();
+	auto stretch = [](auto* table) {
+		// stretching will not work without sorting
+		table->setSortingEnabled(true);
+		table->resizeColumnsToContents();
+	};
+
+	if (type == BreakpointData::ALL) {
+		for (auto* table : tables) {
+			stretch(table);
+		}
+	} else {
+		stretch(tables[type]);
+	}
+}
 
-	if (breakpoints->breakpointCount() != -1) {
-		for (int index = 0; index < breakpoints->breakpointCount(); ++index) {
-			const Breakpoint& bp = breakpoints->getBreakpoint(index);
+std::map<QString, BreakpointData>::iterator
+BreakpointViewer::scanBreakpointData(const Breakpoint& bp)
+{
+	BreakpointData::Type type = bp.type == Breakpoint::BREAKPOINT ? BreakpointData::BREAKPOINT
+	                         : (bp.type == Breakpoint::CONDITION  ? BreakpointData::CONDITION : BreakpointData::WATCHPOINT);
+	if (maps[type].empty()) return maps[type].end();
+
+	// assumed premise: a new bp will never repeat same name again
+	if (auto it = maps[type].find(bp.id); it != maps[type].end() && it->second.index == -1) {
+		// assumed premise 2: if name remains the same, bp is exactly the same.
+		return it;
+	}
 
-			switch (bp.type) {
-			case Breakpoint::BREAKPOINT: {
-				bpRow = createTableRow(BREAKPOINT);
-				setBreakpointStatus(BREAKPOINT, bpRow, true);
+	return maps[type].end();
+}
 
-				bool currentSlot = breakpoints->inCurrentSlot(bp);
-				QBrush color = palette().brush(runState | currentSlot ?
-					QPalette::ColorGroup::Normal : QPalette::ColorGroup::Disabled,
-					QPalette::WindowText);
+void BreakpointViewer::connectBreakpointID(const QString& id, BreakpointData& data)
+{
+	maps[data.type].insert({id, data});
+}
 
-				QTableWidgetItem* item1 = bpTableWidget->item(bpRow, ENABLED);
-				item1->setCheckState(Qt::Checked);
+void BreakpointViewer::sync()
+{
+	// don't reload if self-inflicted update
+	if (selfUpdating) return;
 
-				QTableWidgetItem* item2 = bpTableWidget->item(bpRow, BP_ADDRESS);
-				item2->setForeground(color);
-				item2->setText(hexValue(bp.address, 4));
+	bool empty[BreakpointData::ALL] = {false, false, false};
 
-				QTableWidgetItem* item3 = bpTableWidget->item(bpRow, T_CONDITION);
-				item3->setForeground(color);
-				item3->setText(bp.condition.trimmed());
+	for (int type = 0; type < BreakpointData::ALL; ++type) {
+		if (maps[type].empty()) empty[type] = true;
 
-				QTableWidgetItem* item4 = bpTableWidget->item(bpRow, SLOT);
-				item4->setForeground(color);
-				item4->setText(QString("%1/%2")
-					.arg(bp.ps == -1 ? QChar('X') : QChar('0' + bp.ps))
-					.arg(bp.ss == -1 ? QChar('X') : QChar('0' + bp.ss)));
+		for (auto& it : maps[type]) {
+			it.second.index = -1;
+			it.second.row = -1;
+		}
+	}
```

Comment:
(detail)
The name `it` is a bit misleading: it's not an iterator, it's a pair<key, value>. Often in cases like this the name `p` (for pair) or `e' (for element) is used. (The scope is small enough that a 1-letter variable name is not confusing).

But an alternative that bypasses this naming problem is:
```
for (auto& [id, data] : maps[type]) {
    data.index = -1;
    data.row = -1;
}
```

Unlike QMap, std::map often requires the use of `first` and `second` when accessing elements. Though by using (c++17) structured bindings, this can often be avoided.
I think I see some other places in this patch where this might be useful.

(Just a suggestion, no need to change this).